General Improvements

daxpedda opened the pull request:
I used [trybuild](https://docs.rs/trybuild) to test our error messages. Also turned out there were a lot of bugs, fixed those.

Fixes #2.
Fixes #3.
Fixes #5.
Fixes #7.
Fixes #8.
Fixes #10.
Fixes #18.
Fixes #19.
Fixes #20.

daxpedda commented on src/lib.rs:
```diff
@@ -25,11 +25,28 @@ struct DeriveWhere {
-            .into_iter()
-            .collect();
+        let found_semicolon = input.step(|cursor| {
+            for tree in cursor.token_stream() {
+                if let TokenTree::Punct(punct) = tree {
+                    if punct.as_char() == ';' {
+                        return Ok((true, *cursor));
+                    }
+                }
+            }
+
+            Ok((false, *cursor))
+        })?;
 
-        <Token![;]>::parse(input)?;
+        let bounds = if found_semicolon {
+            let bounds = Punctuated::<TraitBound, Token![,]>::parse_separated_nonempty(input)?
+                .into_iter()
+                .collect();
+            <Token![;]>::parse(input)?;
+
+            bounds
+        } else {
+            Vec::new()
+        };
```
This new implementation is still up for discussion. @ModProg feel free to overwrite it.

ModProg replied:
I would go with fork and advance_to, but let's create an issue about benchmarking this.

daxpedda replied:
Solved in another PR.

daxpedda commented:
Alright, I inverted the syntax and used some custom parsing to improve error messages. This looks much better now. The only thing I would have like is improve error messages on duplicate commas, but the code got so bloated handling that, I reverted everything.

See https://github.com/ModProg/derive-where/issues/3#issuecomment-958028109.

ModProg commented:
> Alright, I inverted the syntax and used some custom parsing to improve error messages. This looks much better now. The only thing I would have like is improve error messages on duplicate commas, but the code got so bloated handling that, I reverted everything.

Well I don't think duplicate commas are a very common mistake, and normally easily spotted by the developer themselves

ModProg commented on tests/ui/invalid_generic.stderr:
```diff
@@ -0,0 +1,5 @@
+error: expected type to bind to, expected one of: `for`, parentheses, `fn`, `unsafe`, `extern`, identifier, `::`, `<`, square brackets, `*`, `&`, `!`, `impl`, `_`, lifetime
+ --> tests/ui/invalid_generic.rs:1:37
+  |
+1 | #[derive_where::derive_where(Clone; where)]
+  |                                     ^^^^^
```
This looks a bit strange, we would only expect a type here, not parentheses ??

daxpedda replied:
I'm not sure what you mean, I don't see any parentheses. I used `where` here as something that can't be parsed into a type.

ModProg replied:
I meant the `expected one of: "..."` but that is probably just the parser returning that. I think we can keep it like that as the error message pointing at the `where` and requiring a Type being pretty clear.

daxpedda replied:
Ah, I see. Good question, I have no clue, but yeah, this error message isn't in my control.

ModProg commented on tests/ui/no_generics.rs:
```diff
@@ -0,0 +1,9 @@
+#[derive_where::derive_where(Clone)]
+struct Test1(u8);
+
+// Make sure this doesn't reach unreachable code as it should fail early for
+// not having any generics.
```
I don't disagree with erroring on these, as the standard derive is always preferable if possible, but I don't think there is unreachable code to reach

daxpedda replied:
This test is making sure that this line is never reached:
https://github.com/ModProg/derive-where/blob/d41f608e029f136e734800767d8a978ea1735c6e/src/lib.rs#L184
It should never be reached because a check happens first to check for any generics. Rust is smart enough to find no generics if they aren't used inside the item.

Maybe the comment needs to be improved?

ModProg replied:
oh ok, no it's ok. 
Even without the unreachable, we just don't want to support it anyways because it doesn't make sense. We can keep it like that.

ModProg commented on tests/ui/union.rs:
```diff
@@ -0,0 +1,9 @@
+#[allow(unused_imports)]
+use core::marker::PhantomData;
+
+#[derive_where::derive_where(Clone)]
+union Test<T> {
+    field: PhantomData<T>,
+}
```
Should we create an issue for unions? To document we could implement them if needed.

daxpedda replied:
I briefly looked into this yesterday, just to make sure why we can't implement them. The issue is that we can't handle unions, because unlike an `enum`, we have no way to check which current variant is active. So really the only trait we could derive for unions is `Copy`, and therefore a special implementation of `Clone`, and `Hash` (I think?).

On that note, standard `derive` only support `Copy`.
So I guess yes, we should create an issue!

daxpedda replied:
See #18.

ModProg commented on src/lib.rs:
```diff
@@ -42,58 +43,63 @@ impl Parse for Generic {
-                    .collect();
-
-                Ok(Self { generics, traits })
+        let mut traits = Vec::new();
+        let mut generics = None;
+
+        // Start parsing traits.
+        while !input.is_empty() {
+            traits.push(Trait::parse(input)?);
+
+            if !input.is_empty() {
+                let fork = input.fork();
+
+                if <Token![;]>::parse(&fork).is_ok() {
+                    input.advance_to(&fork);
+
+                    // If we found a semi-colon, start parsing generics.
+                    if !input.is_empty() {
+                        generics = Some(
+                            Punctuated::<Generic, Token![,]>::parse_terminated(input)?
+                                .into_iter()
+                                .collect(),
+                        );
+                    }
+                } else if let Err(error) = <Token![,]>::parse(input) {
+                    return Err(Error::new(error.span(), "expected `;` or `,"));
+                }
```
:+1: manual parsing and yet almost more readable than my solution

daxpedda replied:
Uff, the first time around when I swapped the syntax without manual parsing it was so beautiful, this looks super-messy to me now :cry:.

ModProg commented on src/lib.rs:
```diff
@@ -554,9 +606,22 @@ impl Trait {
+            Hash => {
+                // Add hashing the variant if this is an `enum`.
+                let discriminant = if variants.is_some() {
+                    let type_ = self.path();
+                    Some(quote! { #type_::hash(&::core::mem::discriminant(self), __state); })
```
`type_` vs `path`, comparing to https://github.com/ModProg/derive-where/pull/15#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R541-R546

daxpedda replied:
Good catch!

daxpedda replied:
Done.

daxpedda commented:
I adjusted the MSRV to 1.36.0 because trybuild doesn't build before that and optional `dev-dependencies` aren't a thing. I also introduced `DERIVE_WHERE_SKIP_UI` environment variable, which disables UI tests, which we can use when testing for MSRV.

ModProg commented:
> I adjusted the MSRV to 1.36.0 because trybuild doesn't build before that

That's ok, I think 1.36 should be good enough.

If necessary, we could do the ui testing in a separate crate to not affect the MSRV. But as there is no known usecase for sub 1.36 let's keep it like it is for now.

daxpedda commented:
I added deriving `Copy` for unions.

daxpedda commented:
I added a crate feature called `safe`. My thoughts were that standard derive uses unsafe everywhere, so we should do the same by default. Switching to a safe implementation is explicit.

I also decided to make the implementation using `transmute` for the `Discriminant` for `Ord` and `PartialOrd` the default one. After digging an hour or so through the code, I determined that this has to be safe. Rust makes an explicit guarantee that single field tuple structs have to be able to be transmuted to the type of their only field. Additionally I added a test that makes sure that the inner type remains the same.

@ModProg let me know if you disagree!

ModProg commented:
> I added a crate feature called `safe`. My thoughts were that standard derive uses unsafe everywhere, so we should do the same by default. Switching to a safe implementation is explicit.
> 
> I also decided to make the implementation using `transmute` for the `Discriminant` for `Ord` and `PartialOrd` the default one. After digging an hour or so through the code, I determined that this has to be safe. Rust makes an explicit guarantee that single field tuple structs have to be able to be transmuted to the type of their only field. Additionally I added a test that makes sure that the inner type remains the same.
> 
> @ModProg let me know if you disagree!

No that makes definitely sense. And if we have CI running these tests we can ensure that this stays the way it is. but I also don't see any reason for them to change it.

daxpedda commented:
Implemented the `nightly` crate feature now. I hope the code isn't too messy.
- `safe` avoids all unsafe
- `nightly` doesn't actually use any unsafe, so it doesn't conflict with `safe`

daxpedda commented:
I went ahead and implemented `Zeroize`.

daxpedda commented:
Remaining:
- Implement `skip`
- Allow on items without generics
- Produce an error when using `default` on non-enums or without a present `Default` trait
- Finish integration tests
- Lastly, improve the code by splitting the logic and hopefully making some modules

@ModProg I know this PR has gotten rather big, it's still reviewable commit-by-commit thought (most of the times 😅). Feel free to tell me to stop and put it in another PR or review or tell me what you think.

ModProg commented:
> @ModProg I know this PR has gotten rather big, it's still reviewable commit-by-commit thought (most of the times sweat_smile). Feel free to tell me to stop and put it in another PR or review or tell me what you think.

It's fine, due to the extensive testing I don't think this will be a problem. 

Only thing we deffinitly need to do is split up the lib.rs 2.6k is just a bit too many lines IMO

daxpedda commented:
> Only thing we deffinitly need to do is split up the lib.rs 2.6k is just a bit too many lines IMO

Absolutely agreed!

ModProg commented:
One idea I would find interesting would be an API that allows 2 (3) different usages:

Simple case:
```rs
#[derive_where::derive_where(Copy, Clone)]
```

Different derives:
```rs
#[derive_where::derive_where]
#[derive_where(Copy, Clone)]
#[derive_where(Debug; T)]
```

Maybe useful to support:
```rs
#[derive_where::derive_where(Copy, Clone)]
#[derive_where(Debug; T)]
```

daxpedda commented:
This is already supported, but differently, because we switched to `proc_macro_derive`.
```rust
#[derive(derive_where::DeriveWhere)]
#[derive_where(Copy, Clone)]
#[derive_where(Debug; T)]
```

ModProg commented:
> This is already supported, but differently, because we switched to `proc_macro_derive`.
> 
> ```rust
> #[derive(derive_where::DeriveWhere)]
> #[derive_where(Copy, Clone)]
> #[derive_where(Debug; T)]
> ```

Yeah I know, that was more an idea of a possible api improvement that we could go for if we want to (serde_with) uses that pattern. 

But that would not change a lot internally as that would be kindof just a wrapper around everything :D

ModProg commented on non-msrv-tests/tests/ui/no_generics.stderr:
```diff
@@ -0,0 +1,15 @@
+3 | | struct Test1(u8);
+  | |_________________^
+
+error: derive-where doesn't support items without generics, `skip` attributes or `enum`s implementing `Default`, as this can already be handled by standard `#[derive(..)]`
```
We should keep in mind, that default might in the future be std: https://doc.rust-lang.org/beta/unstable-book/language-features/derive-default-enum.html

daxpedda replied:
I saw that, we could really use `#[cfg(version(..))]`.

ModProg replied:
yeah, we can make this an issue for now to keep this in mind

ModProg commented on non-msrv-tests/tests/ui/not-zeroize/comma.stderr:
```diff
@@ -0,0 +1,11 @@
+error: unsupported trait syntax, expected one of expected one of Clone, Copy, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd
+ --> tests/ui/not-zeroize/comma.rs:2:22
+  |
+2 | #[derive_where(Clone,; T)]
+  |                      ^
```
Maybe allow trailing commas? As that is the rust default.

daxpedda replied:
We do allow trailing commas, but not when `;` comes afterwards. We might wanna again think about some better syntax, this is all so un-Rusty.

In any case, I can change that.

daxpedda replied:
Had to bump the MSRV to 1.34 too, because 1.33 doesn't support `feature(unrestricted_attribute_tokens)`, which allows things like semi-colons in there:
https://github.com/rust-lang/rust/blob/41301c3b2371365b753c2ad6a74528a38f3815ce/compiler/rustc_feature/src/accepted.rs#L295.

daxpedda replied:
Done.

ModProg replied:
yeah the syntax is indeed unrusty af :D

ModProg commented on src/data/field.rs:
```diff
@@ -0,0 +1,116 @@
+impl Display for Member<'_> {
+	fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+		match self {
+			Member::Named(ident) => f.write_fmt(format_args!("{}", ident)),
```
is there a reason for not using the write! macro?

daxpedda replied:
I actually have no clue, I thought using `format_args` is cheaper, but now I think that's just nonsense. I will look into it.

daxpedda replied:
Actually, what I wrote is exactly what `write!` does under the hood. Fixed!

ModProg replied:
yeah I thought so thats why I was a bit confused and thought there is some msrv going on?

daxpedda commented:
CI is done now too.
